node/src/store.rs

```diff
@@ -1,14 +1,283 @@
+use std::fmt::Debug;
+use std::sync::atomic::{AtomicU64, Ordering};
+
+use async_trait::async_trait;
+use celestia_types::ExtendedHeader;
+use dashmap::mapref::entry::Entry;
+use dashmap::DashMap;
+use tendermint::Hash;
+use thiserror::Error;
+use tracing::{info, instrument};
+
+type Result<T, E = StoreError> = std::result::Result<T, E>;
+
+#[async_trait]
+pub trait Store: Send + Sync + Debug {
+    async fn get_by_hash(&self, hash: &Hash) -> Result<ExtendedHeader>;
+    async fn get_by_height(&self, height: u64) -> Result<ExtendedHeader>;
+
+    async fn height(&self) -> u64;
+    async fn has(&self, hash: &Hash) -> bool;
+    async fn has_at(&self, height: u64) -> bool;
+
+    async fn append_single(&self, header: ExtendedHeader) -> Result<()>;
+
+    async fn append<I: IntoIterator<Item = ExtendedHeader>>(&self, headers: I) -> Result<()>
+    where
+        I: IntoIterator<Item = ExtendedHeader> + Send,
+        <I as IntoIterator>::IntoIter: Send,
+    {
+        let headers = headers.into_iter();
+
+        for (idx, header) in headers.enumerate() {
+            if self.append_single(header).await.is_err() {
+                return Err(StoreError::ContinuousAppendFailedAt(idx));
```

Comment:
An index will not make sense in the error message. Use `height`

Reply:
Or a better idea: In `append_continus` return this:

```
    return NonContinuousAppend(self.get_head_height(), height.value());
```

And here just propagate the error.

Reply:
I think the reference Store interface had it return number of headers successfuly applied, but I like your suggestion better, it simplifies `StoreError` and implementation.